src/object_store/src/object/s3.rs

```diff
@@ -936,85 +979,91 @@ impl Stream for S3ObjectIter {
                     let next_continuation_token = r.next_continuation_token;
                     Ok((more, next_continuation_token, is_truncated))
                 }
-                Err(e) => Err(e),
+                Err(e) => Err(set_error_should_retry::<ListObjectsV2Error>(
+                    config,
+                    e.into(),
+                )),
             }
         };
         self.send_future = Some(Box::pin(f));
         self.poll_next(cx)
     }
 }
 
-type RetryError = Either<
-    SdkError<GetObjectError, aws_smithy_runtime_api::http::Response<SdkBody>>,
-    ByteStreamError,
->;
-
-impl From<RetryError> for ObjectError {
-    fn from(e: RetryError) -> Self {
-        match e {
-            Either::Left(e) => e.into(),
-            Either::Right(e) => e.into(),
-        }
-    }
-}
-
-struct RetryCondition {
-    retry_unknown_service_error: bool,
-    retryable_service_error_codes: Vec<String>,
-}
-
-impl RetryCondition {
-    fn new(config: &S3ObjectStoreConfig) -> Self {
-        Self {
-            retry_unknown_service_error: config.developer.object_store_retry_unknown_service_error
-                || config.retry_unknown_service_error,
-            retryable_service_error_codes: config
-                .developer
-                .object_store_retryable_service_error_codes
-                .clone(),
-        }
-    }
-}
-
-impl tokio_retry::Condition<RetryError> for RetryCondition {
-    fn should_retry(&mut self, err: &RetryError) -> bool {
-        match err {
-            Either::Left(err) => match err {
+fn set_error_should_retry<E>(config: Arc<ObjectStoreConfig>, object_err: ObjectError) -> ObjectError
+where
+    E: ProvideErrorMetadata + Into<BoxError> + Sync + Send + std::error::Error + 'static,
+{
+    let not_found = object_err.is_object_not_found_error();
```

Comment:
nits: I think we can return immediately if `not_found == true` to avoid re-constructing the ObjectStore